src/internal_modules/roc_sndio/target_sndfile/roc_sndio/sndfile_source.cpp

```diff
@@ -0,0 +1,456 @@
+            eof_ = true;
+            break;
+        }
+
+        for (size_t n = 0; n < n_samples; n++) {
+            frame_data[n] = buffer_data[n];
+        }
+
+        frame_data += n_samples;
+        frame_left -= n_samples;
+    }
+
+    if (frame_left == frame.num_samples()) {
+        return false;
+    }
+
+    if (frame_left != 0) {
+        memset(frame_data, 0, frame_left * sizeof(audio::sample_t));
+    }
+
+    return true;
+}
+
+bool SndfileSource::seek_(size_t offset) {
+    roc_panic_if(!valid_);
+
+    if (!sndfile_input_) {
+        roc_panic("sndfile source: seek: non-open input file or device");
+    }
+
+    if (!is_file_) {
+        roc_panic("sndfile source: seek: not a file");
+    }
+
+    roc_log(LogDebug, "sndfile source: resetting position to %lu", (unsigned long)offset);
+
+    sf_count_t err = sf_seek(sndfile_input_, (sf_count_t)offset, SEEK_SET);
+    if (err == -1) {
+        roc_log(LogError,
+                "sndfile source: can't reset position to %lu: an attempt was made to "
+                "seek beyond the start or end of the file",
+                (unsigned long)offset);
+        return false;
+    }
+
+    return true;
+}
+
+bool SndfileSource::setup_names_(const char* driver, const char* path) {
+    if (driver) {
+        if (!driver_name_.assign(driver)) {
+            roc_log(LogError, "sndfile source: can't allocate string");
+            return false;
+        }
+    }
+
+    if (path) {
+        if (!input_name_.assign(path)) {
+            roc_log(LogError, "sndfile source: can't allocate string");
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool SndfileSource::setup_buffer_() {
+    buffer_size_ = sample_spec_.ns_2_samples_overall(frame_length_);
+    if (buffer_size_ == 0) {
+        roc_log(LogError, "sndfile source: buffer size is zero");
+        return false;
+    }
+    if (!buffer_.resize(buffer_size_)) {
+        roc_log(LogError, "sndfile source: can't allocate sample buffer");
+        return false;
+    }
+
+    return true;
+}
+
+bool SndfileSource::open_() {
+    if (sndfile_input_) {
+        roc_panic("sndfile source: already opened");
+    }
+
+    sf_info_in_.format = 0;
+
+    sndfile_input_ = sf_open(input_name_.is_empty() ? NULL : input_name_.c_str(),
+                             SFM_READ, &sf_info_in_);
+    if (!sndfile_input_) {
+        roc_log(LogInfo, "sndfile source: can't open: driver=%s input=%s",
+                driver_name_.c_str(), input_name_.c_str());
+        return false;
+    }
+
+    if (sf_info_in_.channels != (int)sample_spec_.num_channels()) {
+        roc_log(LogError,
+                "sndfile source: can't open: unsupported # of channels: "
+                "expected=%lu actual=%lu",
+                (unsigned long)sample_spec_.num_channels(),
+                (unsigned long)sf_info_in_.channels);
+        return false;
+    }
+
+    if (sample_rate_ != 0) {
+        sf_info_in_.samplerate = sample_rate_;
+    }
```

Comment:
This allows for a check to see if the user has set a sample-rate of their own using --rate. From my understanding sample_rate_, which is initialized by config.sample_spec.sample_rate(), will be 0 if this argument isn't passed. Otherwise, it will force the new sample rate to be what the user entered. Perhaps I am misunderstanding what the --rate argument and the tests for sample_rate_mismatch and sample_rate_auto are for, but without this line, those tests fail.

Reply:
I see... Yes, --rate argument forces specific rate. However, it's intended only for devices, e.g. when you open ALSA or PulseAudio device, it often supports multiple rates natively, and you may want to select one of them.

For files, it doesn't make much sense to force any rate. File has fixed rate. Pipeline will do resampling if pipeline rate differs from file rate.

So I think what we need to do is:

* get sample rate of the file from libsndfile
* report it via sample_spec()
* when opening file, also do a check:
  * if rate in config is zero, any rate is ok, so no problems
  * if rate in config is non-zero, fail if it differs from file rate, because user asks for specific rate, but we can't do it